commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2023 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tiered.storage.commons.chunkindex;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * The chunk index for transformed chunks of variable size.
+ *
+ * <p>The most important quality is that while the original file chunks
+ * are of the same size (apart from maybe the final one), the transformed chunks
+ * have different sizes that must be specified and stored explicitly.
+ * For example, this is a result of compression.
+ *
+ * <p>An example:
+ * <pre>
+ * Original file:        |  Transformed file:
+ * [0-100)   - size 100  |  [0-30)  - size 30
+ * [100-200) - size 100  |  [30-50) - size 20
+ * [200-250) - size 50   |  [50-60) - size 10
+ * </pre>
+ */
+public class VariableSizeChunkIndex extends AbstractChunkIndex {
+    private final List<Integer> transformedChunks;
+
+    // This only a materialization for convenience and performance,
+    // it should not be persisted.
+    private final List<Chunk> chunks;
+
+    public VariableSizeChunkIndex(final int originalChunkSize,
+                                  final int originalFileSize,
+                                  final List<Integer> transformedChunks) {
+        super(originalChunkSize, originalFileSize, transformedChunks.get(transformedChunks.size() - 1));
+        this.transformedChunks = List.copyOf(Objects.requireNonNull(transformedChunks));
+
+        chunks = materializeChunks();
```

Comment:
From Effective Java:
> There are a few more restrictions that a class must obey to allow inheritance. Constructors must not invoke overridable methods, directly or indirectly. If you violate this rule, program failure will result. The superclass constructor runs before the subclass constructor, so the overriding method in the subclass will be invoked before the subclass constructor has run. If the overriding method depends on any initialization performed by the subclass constructor, the method will not behave as expected.

TLDR: we should not call overridable methods from constructors I think.

Reply:
I think this applies to the case where the `AbstractChunkIndex` constructor calls `materializeChunks` and its subsequent calls. However when `FixedSizeChunkIndex` and `VariableSizeChunkIndex` does this, it should be fine as all the necessary state is initialized. I can't find a better way to not repeat the `materializeChunks` code in both classes.

Reply:
Maybe make `materializeChunks` final then? I understand your point and do agree but at the same time I can imagine that somebody might want to patch `FixedSizeChunkIndex` or `VariableSizeChunkIndex` for their use case using inheritance since the classes are not final also. So this is not a problem by itself right now but a potential problem in the future that will be hard to spot if there will be the case.